### PR TITLE
Add  `WandbCallbackHandler` as an available callback

### PR DESCRIPTION
**Pull request by ash0ts**
## Description
This PR adds a Weights & Biases callback to Langchain, which will allow users to easily track the generations of their interaction with LangChain. 

- During the duration of the session with Langchain, callback users have the option to log each action performed by the Module to Weights & Biases, alongside the opportunity to analyze the text using complexity metrics such as Flesch-Kincaid, or an option to visualize the text's structure such as Entity Recognition. Details of stdout are automatically saved to the Weights and Biases platform during the context of the callback.
- At the end of a session with a module, the generations are logged to Weights and Biases, alongside any selected metrics and visualizations. Users have the ability to also log a table containing all the available information for each callback action that was received, similar to a trace.

As there is no formal finish signal to emit depending on the Langchain Module used to provide generations, we treat the usage of these LLMs similar to a user-defined session, where each session is used to collect many generations for a prompt, test out many different prompts, or different configurations for an LLM.
- To serialize the contents of a Langchain module alongside the aforementioned tables to create a contextually tracked model artifact, we add a function to the `WandbCallback` called `flush_tracker` which when provided the langchain module to serialize, and some flags controlling the next state of the model for your sessions, will log the aforementioned artifact to Weights and Biases.

This PR includes the following changes:
- Added `wandb`, `textstat`, `spacy en-core-web-sm` (3.5.0) and `pandas` packages to `pyproject.toml`
- Created a new `WandbCallback` class in `wandb_callback.py` in `langchain/callbacks`
- Exposed this new callback class in the `langchain/callbacks/__.init__.py`

Please let me know if there are any issues or concerns with this PR. Thank you for your time and consideration!

## Recreating Validation
An example script showing 3 different scenarios using Langchain modules with the `WandbCallback` have been provided in this colab: https://colab.research.google.com/drive/1DXH4beT4HFaRKy_Vm4PoxhXVDRf7Ym8L?usp=sharing
An example workspace can be found here: https://wandb.ai/a-sh0ts/langchain_callback_demo?workspace=user-a-sh0ts
- Under `minimal_*` it will show the default workspace which will only log during flush, but still track generations and actions during the session.
    - To recreate the above `minimal` workspace you can run the collar fully as is
- Under `full_*` it will show the maximal workspace with the tracked values during the session, and the expanded generations table with analysis and visualizations
    - To recreate features of the above `full` workspace, set all the appropriate flags to `True` in the callback of the `WandbCallbackHandler`

## Defaults
Default values for WandbCallbackHandler(...):
```
visualize: bool = False,
complexity_metrics: bool = False,
stream_logs: bool = False,
```

Defaults for WandbCallbackHandler.flush_tracker(...):
```
reset: bool = True,
finish: bool = False,
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -51,6 +51,10 @@ pypdf = {version = "^3.4.0", optional = true}
 networkx = {version="^2.6.3", optional = true}
 aleph-alpha-client = {version="^2.15.0", optional = true}
 deeplake = {version = "^3.2.9", optional = true}
+wandb = "^0.13.10"
```
Comment (hwchase17) [inline pyproject.toml:54]: all of these will want to be optional dependencies

File: langchain/callbacks/wandb_callback.py [hunk 2]
```diff
@@ -0,0 +1,713 @@
+try:
```
Comment (hwchase17) [inline langchain/callbacks/wandb_callback.py:1]: the imports in this file need to be made more optional, but im happy to help with that

**Review by hwchase17 — COMMENTED** [review]
an example notebook, or page in the docs/ecosystem folder, would be SUPER helpful for folks trying to get started!

**Comment by ash0ts** [comment]
For brevity this an example of usage

```
wandb_callback = WandbCallbackHandler(
    job_type="inference",
    project="langchain_callback_demo2",
    group=f"minimal_{session_group}",
    name="llm",
    tags=["test"],
)
manager = CallbackManager([StdOutCallbackHandler(), wandb_callback])
llm = OpenAI(temperature=0, callback_manager=manager, verbose=True)
```

where `llm` can be replaced by a `chain` or `agent`

**Comment by ash0ts** [comment]
There is Report here that shows the usage of the callback directly: https://wandb.ai/a-sh0ts/langchain_callback_demo/reports/Prompt-Engineering-LLMs-with-LangChain-and-W-B--VmlldzozNjk1NTUw#👋-how-to-build-a-callback-in-langchain-for-better-prompt-engineering

**Pull request merged**